pig/src/main/java/com/netflix/iceberg/pig/PigParquetReader.java

```diff
@@ -0,0 +1,523 @@
+    }
+  }
+
+  private static class LongDecimalReader extends PrimitiveReader<BigDecimal> {
+    private final int scale;
+
+    LongDecimalReader(ColumnDescriptor desc, int scale) {
+      super(desc);
+      this.scale = scale;
+    }
+
+    @Override
+    public BigDecimal read(BigDecimal reuse) {
+      return BigDecimal.valueOf(column.nextLong(), scale);
+    }
+  }
+
+  private static class MapReader<K, V> extends RepeatedKeyValueReader<Map<K, V>, Map<K, V>, K, V> {
+    ReusableEntry<K, V> nullEntry = new ReusableEntry<>();
+
+    MapReader(int definitionLevel, int repetitionLevel,
+              ParquetValueReader<K> keyReader, ParquetValueReader<V> valueReader) {
+      super(definitionLevel, repetitionLevel, keyReader, valueReader);
+    }
+
+    @Override
+    protected Map<K, V> newMapData(Map<K, V> reuse) {
+      return new LinkedHashMap<>();
+    }
+
+    @Override
+    protected Map.Entry<K, V> getPair(Map<K, V> reuse) {
+      return nullEntry;
+    }
+
+    @Override
+    protected void addPair(Map<K, V> map, K key, V value) {
+      map.put(key, value);
+    }
+
+    @Override
+    protected Map<K, V> buildMap(Map<K, V> map) {
+      return map;
+    }
+  }
+
+  private static class ArrayReader<T> extends RepeatedReader<DataBag, DataBag, T> {
+    private final BagFactory BF = BagFactory.getInstance();
+    private final TupleFactory TF = TupleFactory.getInstance();
+
+    ArrayReader(int definitionLevel, int repetitionLevel, ParquetValueReader<T> reader) {
+      super(definitionLevel, repetitionLevel, reader);
+    }
+
+    @Override
+    protected DataBag newListData(DataBag reuse) {
+      return BF.newDefaultBag();
+    }
+
+    @Override
+    protected T getElement(DataBag list) {
+      return null;
+    }
+
+    @Override
+    protected void addElement(DataBag bag, T element) {
+      bag.add(TF.newTuple(element));
+    }
+
+    @Override
+    protected DataBag buildList(DataBag bag) {
+      return bag;
+    }
+  }
+
+  private static class TupleReader extends StructReader<Tuple, Tuple> {
+    private static final TupleFactory TF = TupleFactory.getInstance();
+    private final Map<Integer, Object> partitionValues;
+    private final int columns;
+
+    protected TupleReader(List<Type> types, List<ParquetValueReader<?>> readers, Map<Integer, Object> partitionValues) {
+      super(types, readers);
+
+      this.partitionValues = partitionValues;
+      this.columns = types.size() + partitionValues.size();
+    }
+
+    @Override
+    protected Tuple newStructData(Tuple reuse) {
+      return TF.newTuple(columns);
+    }
+
+    @Override
+    protected Object getField(Tuple tuple, int pos) {
+      return null;
+    }
+
+    @Override
+    protected Tuple buildStruct(Tuple tuple) {
+      for (Map.Entry<Integer, Object> e : partitionValues.entrySet()) {
+        try {
+          tuple.set(e.getKey(), e.getValue());
+        } catch (ExecException ex) {
+          throw new RuntimeException("", ex);
```

Comment:
Can you update the error message to include `e.getKey()`? That may be helpful if anyone ever hits this.